Allow custom parsing of EntitySelector

Mathe172 opened the pull request:
## What does it do
This PR adds the possibility to override the way the EntitySelector is parsed. It does so by routing calls to the three affected methods (`EntitiySelector.matchEntities(ICommandSender, String, Class)`, `.hasArguments(String)` and `.matchesMultiplePlayers(String)`) over an instance of `SelectorHandler`. The default implementation of this class just calls the old implementation (their names are now `*Default`).

~~Note: We're not sure whether `ForgeHooks` is the best place for the proxy, if there's anything better, we will obviously change that~~
~~Note 2: Currently, the patch changes 6 lines of `EntitySelector`. This could be reduced to 3 by removing the line break between the two, but this is probably worse than 6 lines.~~

## Why this way
### Why these methods
Overriding these three methods allows you to change the behavior of all public methods of `EntitySelector` (except `getScoreMap`, where overriding doesn't make sense). `matchesMultiplePlayers` and `hasArguments` need to be overridden too, if one wants to get consistent behaviour.

### Why not as an event
The issue with using events is that either you have to listen for three events (not reflecting that they belong together) or to a single compound event for all three methods (which is even messier).

## Why at all
Doing it this way allows you to affect all places where selectors are used at once: Commands, IChatComponents and CommandResultStats. While it would be possible to do it for commands using the `CommandEvent` or by overriding the `CommandHandler`, it would be a lot harder, and you would have to use another trick for the other two cases, not to mention that you would never catch if someone used a selector in something entirely different.

For simple additions of other parameters, it would be easier to just provide a hook allowing you to add additional `Predicate`s inside the `matchEntities` method. However, it is not possible to use this approach to introduce a more general syntax. An example would be something like
```
@e[r=$score[@p[rm=5,r=10],radius]]
```
This would read out the score `radius` from the player between `5` and `10` blocks away, and use the resulting value for the `r` parameter of the `@e` selector. The issue here is that you don't know the syntax of the values (the `@e[<key1>=<value1>,...]` structure should remain the same) in general (since they could be literally anything). This means that it can't be fixed by simply changing the RegEx or so (even if an extremely messy recursive RegEx (which is not even part of Java) could potentially do it for this particular example), so you really have to parse the selector yourself.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3353) <br/>All committers have signed the CLA.

liach commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -125,6 +127,30 @@
+     * Allows you to override the behavior of EnitySelector by setting this proxy to a custom implementation
+     * Note: You should generally override all three methods to produce consistent results
+     */
+    public static EntitySelectorProxy entitySelectorProxy = new EntitySelectorProxy();
```
I suggest to add getter and setter for this. Report when this is set twice by different mods.

LexManos replied:
Yes, getters/setters is the way to go so we have some control over who goes where.

LexManos commented on patches/minecraft/net/minecraft/command/EntitySelector.java.patch:
```diff
@@ -12,3 +12,30 @@
+     }
+ 
+     public static boolean func_82377_a(String p_82377_0_)
++    {return net.minecraftforge.common.ForgeHooks.entitySelectorProxy.matchesMultiplePlayers(p_82377_0_);}
```
you dont need to compact it THIS much brances on new lines is fine.
The patches just need to be clear what they are doing at a quick glance.

LexManos commented:
I am reluctant to give just a sloppy 'replace it because you feel like it' solution.
Are there any other justifications you can provide besides just wanting recursive syntax? Something that would justify making it a replacement like this and not simply a better implementation?
And no ForgeHooks is NOT the place to stick interfaces that you expect modders to implement.... We have packages for a reason find a good one. There is no need for it to be a inner class.

Another one of my major concerns is that for the vanilla world, commands are all they have.
There needs to be something in place to the best of our ability to make sure that the extension of the syntax does not conflict, or break vanilla functionality. Again, another reason why i'm against the sloppy replacement system you have here.

You keep talking about regex, regex isn't needed here it's just how they decided to implement the vanilla version.
There are a million different ways you can tokenize a string, 'recursive regex' doesnt have to be one of them.

Mathe172 commented:
To give you a bit of context: We're currently working on a mod that completely rewrites the command system - it allows for things like variables, nesting and arbitrarily general syntax for parameters. <sup>(don't look at the repo as it is in very bad shape and not really relevant for this PR.  It will be updated once we change to forge, since the project structure is currently a huge mess)</sup>. For consistency reasons, we would like to be able to provide the same functionality in all selectors. The problem with a more controlled framework is that it would be extremely artificial and overloaded: Issues include:
- The parameters can't be split by the framework, this has to be done by the mod
- To allow for variables, you would need some kind of parser state through which arbitrary information can be exchanged by the mod

While it would be relatively easy to implement this, we feel like it has no place in Forge itself. Additionally it would be very specific to our mod, and would still not be very handy for what we intend.

We do realize that the current state of the PR seems to suggest that you will not be able to simply add parameters (which is probably what most people would like to do). Therefore, we would suggest the following:
- Add some kind of event, which can be used to add additional predicates to selectors without changing the syntax (along the lines of #3356)
- Route all methods over some kind of SelectorHandler (i.e. what the PR currently does), similar to the `CommandHandler` <sup>(we know that you shouldn't normally override that either, but we think it is by far the cleanest way in our case, both to implement our mod and to maintain compatibility)</sup>
- Encourage modders to use the event as an easier and more compatible way to add functionality to selectors, whenever possible.

This way, mods can simply add additional parameters. But mods who want to can still change the complete behavior - obviously, they would have to take care of compatibility. Another advantage of this is that mods that do chose to override the SelectorHandler can implement a much cleaner framework suited for their purposes, which could even be used by other mods if they want to provide that kind of functionality.

Does that seem like a viable solution to you?

Commit:
Rerouted EntitySelector methods over SelectorHandler

Added SelectorHandler
Rerouted EntitySelector.matchEntities(ICommandSender, String, Class), .hasArguments(String) and .matchesMultiplePlayers(String) over SelectorHandler (this allows one to change the behavior of the selector parsing process)
The default implementation of SelectorHandler uses the old implementations, which are renamed to *Default(...)
Adjusted forge.exc to correctly obfuscate parameters of the new *Default(...) methods

Mathe172 commented:
We updated the PR to incorporate the suggestions (for getters/setters etc.) and added more documentation that refers you to `EntitySelectorEvent` for simpler use cases.

shadowfacts commented:
Hasn't this been superseded by #3356?

Mathe172 commented:
No, it has not. If you read our last comment (or the one before), you'll notice that this is intended as a more general approach. As mentioned, we added a comment referring you to #3356 for more "straightforward" changes

mezz commented:
Closing this for the same reason as the 1.11 version, conversation can continue there.
